### PR TITLE
Remove jQuery, replace with Stimulus

### PR DESCRIPTION
**Pull request by nicksayre**
### Description of Change

1. Remove jQuery CDN reference
2. Convert existing jQuery code to use Stimulus

### Related Issue
Closes https://github.com/adhocteam/nytimes-library/issues/36

### Motivation and Context

Removing jQuery is motivated by a desire for a more modern tech stack. We went with Stimulus due to our perception that it doesn't take us too far from the upstream project. It preserves the HTML, keeps full-page renders, routes, etc.

### CODE REVIEW

File: custom/layouts/partials/usefulnessModal.ejs [hunk 1]
```diff
@@ -34,10 +34,17 @@
           <header>
             What is the reason for your feedback?
           </header>
+<<<<<<< HEAD
```
Comment (gunsch) [inline custom/layouts/partials/usefulnessModal.ejs:37]: merge markers

File: public/scripts/main.js [hunk 2]
```diff
@@ -1,213 +1,99 @@
...
+//     console.log('items', items)
+
+//     $('ul.team-cat-list').prepend(items)
+//     $('ul.team-cat-list li:gt(' + (expectedLength - 1) + ')').detach()
+//   })
+
+//   /*
+//     This code swaps "Favorite Docs" into the "Useful Docs" panel if you have at least three favorites.
+//     We decided that we'll disable for v1 but perhaps incorporate after initial launch.
+
```
Comment (gunsch) [inline public/scripts/main.js:55]: ohhhh this seems kinda useful at some point

File: public/scripts/main.js [hunk 3]
```diff
@@ -1,213 +1,99 @@
...
+// function refreshHistory(localStorageKey, type, cb) {
+//   console.log('refreshHistory')
+//   $.ajax('/reading-history/' + type + '.json?limit=5', {
+//     success: function(data) {
+//       localStorage.setItem(localStorageKey, JSON.stringify({ ts: new Date(), history: data }))
+//       if(cb) { return cb(data) }
+//     }
+//   })
+// }
```
Comment (gunsch) [inline public/scripts/main.js:99]: hmm are we not pulling forward the history refresh part? (though tbh I'm not 100% sure what this does)

File: scripts/controllers/edit_button_controller.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+  handleClick() {
+    fetch(
+      window.location.pathname + '?edit=1',
+      {
+        method: 'GET',
+      }
+    )
+  }
+
```
Comment (gunsch) [inline scripts/controllers/edit_button_controller.ts:13]: this seems... kinda overkill for just `href="?edit=1"` ?

File: scripts/controllers/search_controller.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    var lastFilenameFetch = this.#getFilenameStorage()
+    var now = new Date().getTime()
+    if (!lastFilenameFetch || now - lastFilenameFetch.lastFetched > 600000) { // 10 min
+      fetch(
+        '/filename-listing.json',
+        {
+          method: 'GET',
+        }
+      )
+      .then((response) =>{
```
Comment (gunsch) [inline scripts/controllers/search_controller.ts:26]:
```suggestion
      .then((response) => {
```

File: scripts/controllers/search_controller.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  static targets = [ "form", "list", "searchInput" ]
+
+  keypress : boolean
+  formTarget: HTMLFormElement
+  searchInputTarget: HTMLInputElement
+  listTarget: HTMLElement
+
+
+  connect() {
```
Comment (gunsch) [inline scripts/controllers/search_controller.ts:13]:
```suggestion
  listTarget: HTMLElement

  connect() {
```

File: scripts/controllers/search_controller.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+  #filenameMatcher(query: string, callback: Function) {
+    const substrRegex = new RegExp(query, 'i')
+    const filenames = this.#getFilenameStorage().filenames
+    callback(filenames.filter((item: string) => {
+      return substrRegex.test(item)
+    }))
+  }
+
+  #getFilenameStorage = ()=>  {
```
Comment (gunsch) [inline scripts/controllers/search_controller.ts:72]:
```suggestion
  #getFilenameStorage = () =>  {
```

File: scripts/controllers/search_controller.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  handleInput() {
+    if (this.keypress === false) {
+      this.formTarget.insertAdjacentHTML('beforeend', '<input type="hidden" name="autocomplete" value="1"/>');
+      this.formTarget.submit()
+    }
+    this.keypress = false
+  }
+
+  #filenameMatcher(query: string, callback: Function) {
+    const substrRegex = new RegExp(query, 'i')
```
Comment (gunsch) [inline scripts/controllers/search_controller.ts:65]: I've never seen a convention like this (`#` to start... private methods?) is that a stimulus thing?

File: scripts/controllers/user_tools_controller.ts [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+      var userId = (data || {}).analyticsUserId;
+      if (userId) {
+        window['ga']('set', 'userId', userId)
+
+        // We don't use this feature
+        // if(window.location.pathname === '/') {
+        //   $(document).ready(function() {
+        //     personalizeHomepage(userId)
+        //   })
+        // }
```
Comment (gunsch) [inline scripts/controllers/user_tools_controller.ts:36]: if you want to keep this commented-out blob here, probably move `personalizeHomepage` into this file too

(or just delete them both. they're always in git history if we need them)

File: scripts/controllers/user_tools_controller.ts [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+          '<a href="' + path + '">',
+            '<p class="docs-title">' + doc.prettyName + '</p>',
+            '<p class="docs-attr">',
+              '<span class="docs-folder">' + folder + '</span>',
+              '<span class="timestamp">(' + viewedDocument.lastViewed + ')</span>',
+            '</p>',
+          '</a>',
+        '</li>'
+        // use .join() to turn to html string
+        ].join('')
```
Comment (gunsch) [inline scripts/controllers/user_tools_controller.ts:96]: maybe use multiline literals here instead? (here and throughout PR)

File: scripts/controllers/user_tools_controller.ts [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    });
+
+    var className = elementAttributes.name.toLowerCase().replace(' ', '-') + '-content';
+
+    var fullSection = [
+      "<h3>" + elementAttributes.name + "</h3>",
+      "<ul class='" + className + "'>" + items.join('') + "</ul>"
+    ].join('');
```
Comment (gunsch) [inline scripts/controllers/user_tools_controller.ts:104]: multiline literal

File: styles/partials/core/_categories.scss [hunk 12]
```diff
@@ -97,8 +97,8 @@
 
   .g-left-panel {
     float: left;
-    position: relative;
-    top: 8px;
+    position: sticky;
+    top:60px;
```
Comment (gunsch) [inline styles/partials/core/_categories.scss:101]:
```suggestion
    top: 60px;
```

File: scripts/controllers/edit_button_controller.ts (hunk 4 above)
Comment (nicksayre) [reply scripts/controllers/edit_button_controller.ts:13]: Haha, yes it does. I may have I gotten so caught up porting that I didn't analyze the original code's usefulness. Here's the original. 

```
  $(document).on('click', '#edit-button', function () {
    $.ajax({
      method: 'GET',
      url: window.location.pathname,
      data: {
        edit: 1, // purge and keep out of cache temporarily
      }
    }) // no callback on this.
  })
```

I don't really get the code comment: Purge what? Why keep out of cache? Why temporarily? I think we can just use a regular link, as you suggested, unless this code snippet changes your mind.

File: scripts/controllers/search_controller.ts (hunk 8 above)
Comment (gunsch) [reply scripts/controllers/search_controller.ts:65]: whoa! TIL

File: scripts/controllers/edit_button_controller.ts (hunk 4 above)
Comment (gunsch) [reply scripts/controllers/edit_button_controller.ts:13]: I did some git history digging, looks like it's 
https://github.com/nytimes/library/commit/bcb4cd0c19de711b54b94a79b6ab57fd71bd623e#diff-04c97f3a78a1f0cbdcd4754ca986bbafb55bb3072dac3b3df1f2dcbccbdb7f5cR19

seems like it's an app functionality triggered on the backend by simply setting the `edit=1` query parameter. so yeah just a link should be fine!

File: public/scripts/main.js (hunk 3 above)
Comment (nicksayre) [reply public/scripts/main.js:99]: I think all this is superfluous unless we want to enable the `teams` module at some point. Currently we have removed it from`/custom/strings.yaml`. `refreshHistory` is only called by `fetchHistory` which is only called by `personalizeHomepage`, so all those functions are doing nothing at the moment. I can convert it all to JS / Stimulus though if we think it's worth it!

Comment (gunsch) [reply public/scripts/main.js]: ohhh interesting. at a first glance I thought maybe there was a "personal history" component here separate from the "teams" module. nah it's fine to remove

**Review by gunsch — APPROVED** [review]

**Pull request merged**